### PR TITLE
All variable import

### PR DESCRIPTION
**Pull request by nickgraz**
- Create a new "all variable" import
- Mostly a copy of the manifest importer, except everything related to sections is removed
- Adds an admin endpoint to receive the CSV
- Adds a form in the document producer admin for the all variable import upload
- Adds the following new fields used in the "all variables" csv import
```
deliverable_id
deliverable_question
dependency_variable_stable_id
dependency_condition_id
dependency_value
internal_only
```
- Adds a new enum table for "dependency condition" and some basic conditions
- Variable store is modified to have a fetch context for variable manifest imports as well as all variable imports
- Within a single variable, there is a "dependency configuration" which is dictated by the 3 `dependency_*` fields. If any of them are filled out, they must all be filled out.

- [x] Add tests
- [x] Determine where to put the section variable manifest validation - **I split up the two imports so they each have their own validation now, section validation will stay in the variable manifest importer**
- [ ] Investigate failed tests after removing table column check in `fetchManifestVariables` - [thread](https://github.com/terraware/terraware-server/pull/2071/files/bf383267c23d4cd83175a6f4320169fb46e9211c#r1647848943)

### CODE REVIEW

**Comment by nickgraz** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2071?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2071** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2071?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2033** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2033?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 14 other dependent PRs ([#2029](https://github.com/terraware/terraware-server/pull/2029) <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2029?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#2064](https://github.com/terraware/terraware-server/pull/2064) <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2064?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#2065](https://github.com/terraware/terraware-server/pull/2065) <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/2065?utm_source=stack-comment-icon" target="_blank">Graphite</a> and 11 others)
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @nickgraz and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/main/kotlin/com/terraformation/backend/admin/AdminDocumentProducerController.kt [hunk 1]
```diff
@@ -61,6 +63,39 @@ class AdminDocumentProducerController(
...
   }
 
+  @Operation(
+      summary = "Upload a variable manifest.",
+      description = "The uploaded file must be in CSV format.",
+      externalDocs =
+          ExternalDocumentation(
+              description = "Current Google sheet",
+              url =
+                  "https://docs.google.com/spreadsheets/d/16ybKFseSwFubxZ_i6ENxsF2XT0RqajhdumefJOoHQo8/edit#gid=1050189377"))
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/admin/AdminDocumentProducerController.kt]: We should make sure the permissions on this document are locked down tight since terraware-server is a public repo.

Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/admin/AdminDocumentProducerController.kt]: It is currently only accessible to people with a Terraformation email

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/admin/AdminDocumentProducerController.kt]: Looks that way to me too; we'll just need to make sure we're careful with sharing that doc from now on since we don't have security-by-obscurity as an added hurdle.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 2]
```diff
@@ -85,7 +93,23 @@ class VariableStore(
...
+   * Returns information about a variable. Eagerly fetches other variables that are related to the
+   * requested variable, e.g., table columns.
+   *
+   * @throws VariableNotFoundException The variable didn't exist.
+   * @throws VariableIncompleteException The variable didn't have all the information required for
+   *   its type.
+   * @throws CircularReferenceException There is a cycle in the graph of the variable and its
+   *   children.
+   */
+  fun fetchVariableForStableId(stableId: String, variableId: VariableId): Variable {
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Why are both the stable and variable IDs needed here? Can't you look up the stable ID given just the variable ID?

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 3]
```diff
@@ -255,7 +298,128 @@ class VariableStore(
...
+                .from(VARIABLE_SECTION_RECOMMENDATIONS)
+                .where(SECTION_VARIABLE_ID.eq(base.id))
+                .and(VARIABLE_MANIFEST_ID.eq(manifestId))
+                .fetch(RECOMMENDED_VARIABLE_ID.asNonNullable())
+          }
+
+      return SectionVariable(base, sectionRow.renderHeading!!, children, recommends)
+    }
+
+    protected fun fetchTable(base: BaseVariableProperties): TableVariable {
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Should this be in the other fetch class? Tables should be entirely in the all-variables sheet, I believe.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/AllVariableCsvVariable.kt [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    /** Column 12/L - Header? (for table columns) */
+    val isHeader: Boolean,
+    /** Column 13/M - Notes */
+    val notes: String?,
+    /** Column 14/N - Deliverable ID */
+    val deliverableId: DeliverableId?,
+    /** Column 15/O - Deliverable Question */
+    val deliverableQuestion: String?,
+    /** Column 16/P - Dependency - Variable ID */
+    val dependencyVariableId: VariableId?,
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/AllVariableCsvVariable.kt]: Wouldn't this need to be a stable ID, not a variable ID? If you're adding a new variable and another new variable that depends on it, there won't be a variable ID for the first one yet.

**Review by sgrimm — COMMENTED** [review]
Not done looking yet, but a few initial comments.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/AllVariableCsvVariable.kt (hunk 4 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/AllVariableCsvVariable.kt]: Good point, will fix it up

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 3 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: That is correct, thanks for the catch

Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Looks like I had it in both places, will remove from this spot

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 2 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: That is true, I was trying to repurpose the code used for the manifest import, but it seems like it is unnecessary as you point out

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 5]
```diff
@@ -67,6 +67,14 @@ class VariableStore(
    */
   private val variables = ConcurrentHashMap<Pair<VariableManifestId, VariableId>, Variable>()
 
+  /**
+   * Cache of information about variables by their stable ID. Each stable ID will point to the
+   * latest version of the variable for that ID. This just holds variable definitions, not values!
+   * There is currently no mechanism to limit the size of this cache; the assumption for now is that
+   * all the variables in the system will easily fit in memory.
+   */
+  private val stableIdVariables = ConcurrentHashMap<Pair<String, VariableId>, Variable>()
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: I suspect this will be changing or going away based on the discussion on `fetchVariableForStableId`, but if we're keeping it in some form, we'll need to be careful  about cache invalidation. The variables cache (by variable ID) doesn't need invalidation because variables are immutable, but the variable a stable ID points to can change over time.

It's not enough to just remove the entry from this map as part of the all-variables import, because in staging/prod, we're running two terraware-server instances each of which will have its own in-memory cache. Just removing the entry from the map on whichever instance is handling the import request would leave a stale entry on the other instance.

My recommendation for now would be to avoid trying to cache which variable a stable ID points to. If it turns out to be a performance issue, we can look at adding a cache but I'd rather avoid having to do distributed cache management unless we really need to.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 6]
```diff
@@ -114,6 +138,28 @@ class VariableStore(
...
 
+  /**
+   * Fetches the list of non-section variables, using the latest version of each based on their
+   * stable ID. Table columns are not included in the top-level list, but rather in the containing
+   * variables' lists of children.
+   */
+  fun fetchAllNonSectionVariables(): List<Variable> {
+    return with(VARIABLES) {
+      dslContext
+          .select(ID, STABLE_ID)
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: There needs to be an aggregation function on `ID` for the `groupBy` to be valid.

Try `select(DSL.max(ID), STABLE_ID)` and remove the `orderBy`.

**Review by sgrimm — COMMENTED** [review]
A couple more comments; will wait for the next revision before looking at it in more detail.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 5 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: I removed the stable ID from this, so now it is just variable ID -> variable. Let me know if that is alright.

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Yeah, that solves the invalidation problem.

I ended up having to modify this in #2075 too, but I took a somewhat different approach. We should discuss which way we want to go.

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: My thought was that for a given variable, it should either always have a manifest ID (if it's a section) or never have one (if it's any other kind of variable). So I kept it in one map and made the manifest ID nullable in the key.

Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: I see, that seems like a good way to go. I will update my PR so it is closer to yours.

**Comment by nickgraz** [comment]
@sgrimm I think this is ready for a re-review

File: src/main/kotlin/com/terraformation/backend/accelerator/db/DeliverableStore.kt [hunk 7]
```diff
@@ -28,6 +28,14 @@ import org.jooq.impl.DSL
 class DeliverableStore(
     private val dslContext: DSLContext,
 ) {
+  fun deliverableIdExists(id: DeliverableId): Boolean {
+    requirePermissions { readAllDeliverables() }
+
+    return with(DELIVERABLES) {
+      dslContext.select(ID).from(this).where(ID.eq(id)).fetchOne()?.value1()?.let { true } ?: false
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/accelerator/db/DeliverableStore.kt]: jOOQ has a convenience method for this:
```kotlin
dslContext.fetchExists(DELIVERABLES, DELIVERABLES.ID.eq(id))
```

File: src/main/kotlin/com/terraformation/backend/admin/AdminDocumentProducerController.kt (hunk 1 above)
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/admin/AdminDocumentProducerController.kt]: It's harmless to have, but we don't include `/admin` endpoints in the OpenAPI schema since they're not intended for use outside the admin UI, so this documentation will be invisible.

It might be convenient to put a link to the spreadsheet in the admin UI itself so people can navigate to it from the admin UI easily. (I should have mentioned the OpenAPI thing on the earlier PR that added the endpoint below this one.)

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 8]
```diff
@@ -87,10 +87,24 @@ class VariableStore(
    * @throws CircularReferenceException There is a cycle in the graph of the variable and its
    *   children.
    */
-  fun fetchVariable(variableId: VariableId, manifestId: VariableManifestId? = null): Variable {
+  fun fetchVariableForManifest(manifestId: VariableManifestId, variableId: VariableId): Variable {
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Curious why this is being split into two methods. In #2075 I had changed it to make the manifest ID an optional parameter (and put it after the variable ID in the argument list) such that you could do `fetchVariable(variableId, manifestId)` to get the behavior of this method, or `fetchVariable(variableId)` to get the behavior of the other method.

It's not a huge deal either way, just curious why you didn't like that design.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 9]
```diff
@@ -187,11 +222,10 @@ class VariableStore(
    * Logic for recursively fetching a variable and the variables it's related to. Variable fetching
    * is stateful because we want to detect cycles.
    */
-  private inner class FetchContext(private val manifestId: VariableManifestId?) {
+  private inner class FetchContext(val manifestId: VariableManifestId? = null) {
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Why the change from private to public here?

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 10]
```diff
@@ -113,7 +127,28 @@ class VariableStore(
                   .and(VARIABLE_SECTIONS.PARENT_VARIABLE_ID.isNotNull))
           .orderBy(POSITION)
           .fetch(VARIABLE_ID.asNonNullable())
-          .map { fetchVariable(it, manifestId) }
+          .map { fetchVariableForManifest(manifestId, it) }
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: The not-exists check on `VARIABLE_TABLE_COLUMNS` can probably go away since sections can't ever be table columns.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 11]
```diff
@@ -203,8 +237,19 @@ class VariableStore(
...
+            } else {
+              null
+            }
+        val recommendedBy =
+            if (manifestId != null) {
+              fetchRecommendedBy(variableId)
+            } else {
+              emptyList()
+            }
+
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Will this now always be an empty list? A section will never be recommended by another section.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 12]
```diff
@@ -276,15 +317,49 @@ class VariableStore(
...
+          base.manifestId
+              ?: with(VARIABLE_MANIFEST_ENTRIES) {
+                dslContext
+                    .select(VARIABLE_MANIFEST_ID)
+                    .from(this)
+                    .where(VARIABLE_ID.eq(base.id))
+                    .fetchOne(VARIABLE_MANIFEST_ID.asNonNullable())
+                    ?: throw VariableManifestNotFoundForVariableException(base.id)
+              }
+
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt:315]: Under what circumstances would the query happen here? Either you're fetching non-manifest variables in which case you won't be fetching any sections, or you're fetching manifest variables in which case the manifest ID is already in the fetch context.

That aside, a couple issues here:

This variable name shadows the one in `FetchContext` which is probably going to cause confusion down the road. Better to pick a distinct name.

If you upload a revised version of a manifest, any sections that didn't change will have two `variable_manifest_entries` rows and the `fetchOne()` here will fail because the query returns more than one result.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt [hunk 13]
```diff
@@ -263,7 +304,7 @@ class VariableStore(
       }
     }
 
-    private fun fetchNumber(base: BaseVariableProperties): NumberVariable {
+    protected fun fetchNumber(base: BaseVariableProperties): NumberVariable {
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Why the change from `private` to `protected` on the fetch methods? There don't appear to be any subclasses of `FetchContext` that would need to inherit these methods.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+
+    existingDeliverableIds.add(deliverableId)
+  }
+
+  private fun validateDependencyVariableStableId(value: String?, field: String) {
+    if (value.isNullOrBlank()) {
+      return
+    }
+
+    if (value !in existingStableIds) {
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt:110]: Not a problem, but noting that this will mean you have to have the dependencies listed in the CSV before the dependent variables. It would be weird to list them the other way around anyway, though, and this has the nice side effect of preventing cycles. Just something to make sure the people editing the variable list are aware of.

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt:110]: Though maybe this could use a sanity check to detect a variable depending on itself (since at this point the variable's stable ID will have been added to `existingStableIds` already).

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+    if (!configSupplied) {
+      return
+    }
+
+    dependencyConfig.forEach { (value, message) ->
+      if (value == null) {
+        addError(null, null, message)
+      }
+    }
+  }
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt]: Obviously this isn't a hot code path, but constructing all three error messages for every row in the spreadsheet feels inefficient to me. How about something like this?
```kotlin
val columns =
    listOf(
        VARIABLE_CSV_COLUMN_INDEX_DEPENDENCY_VARIABLE_STABLE_ID,
        VARIABLE_CSV_COLUMN_INDEX_DEPENDENCY_CONDITION,
        VARIABLE_CSV_COLUMN_INDEX_DEPENDENCY_VALUE)
val missingColumns = columns.filter { values[it] == null }

if (missingColumns.isNotEmpty() && missingColumns.size != columns.size) {
    missingColumns.forEach { index ->
        val columnName = getColumnName(index)
        addError(columnName, null, messages.variablesCsvDependencyConfigIncomplete(columnName))
    }
}
```
One other thing I did there was pass the column name to `addError()`. So it should no longer be necessary to also include the column name in the actual error message, that is, `Messages.variablesCsvDependencyConfigIncomplete()` can stop taking an argument.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableImporter.kt [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+          // Import all non tables, non sections
+          importAllVariableCsvVariables(
+              csvVariables = csvVariables,
+              ignoreVariableTypes =
+                  listOf(HierarchicalVariableType.Table, HierarchicalVariableType.Section))
+
+          // Import all sections
+          importAllVariableCsvVariables(
+              csvVariables = csvVariables, importVariableType = HierarchicalVariableType.Section)
+
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableImporter.kt]: Will there ever be sections to import?

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/AllVariableCsvVariable.kt [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+
+    fun create(input: String): AllVariableCsvVariableType =
+        byValue[input] ?: throw IllegalArgumentException(LOCALIZED_ERROR_KEY_UNKNOWN_DATA_TYPE)
+  }
+}
+
+enum class HierarchicalVariableType(val value: String) {
+  Table("Table"),
+  Section("Section")
+}
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/AllVariableCsvVariable.kt]: Maybe this should move to its own top-level file since it's used by both the all-variables and the manifest importer.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableImporter.kt [hunk 18]
```diff
@@ -0,0 +1,464 @@
...
+        section: SectionVariable
+    ): Boolean {
+      val csvSubsections = csvVariablesByParentPath[csvVariable.variablePath] ?: emptyList()
+
+      return csvSubsections.size == section.children.size &&
+          section.children.zip(csvSubsections).all { (subsection, csvSubsection) ->
+            canReuseExistingVariable(csvSubsection, subsection)
+          }
+    }
+
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableImporter.kt]: Is this needed?

File: src/main/kotlin/com/terraformation/backend/documentproducer/model/Variable.kt [hunk 19]
```diff
@@ -441,15 +416,22 @@ data class SectionVariable(
         throw VariableValueInvalidException(id, "Display style is required for variable injections")
       }
 
-      val usedVariableId = value.value.usedVariableId
+      // TODO validate this somewhere else
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/documentproducer/model/Variable.kt]: What was wrong with the existing validation setup?

File: src/test/kotlin/com/terraformation/backend/documentproducer/db/VariableImporterTest.kt [hunk 20]
```diff
@@ -0,0 +1,821 @@
...
+    @Test
+    fun `detects duplicate stable IDs`() {
+      val testCsv =
+          header +
+              "\nName X,Duplicate ID,,Number,Yes,,,,,,,,,,,,,," +
+              "\nName Y,Duplicate ID,,Number,Yes,,,,,,,,,,,,,,"
+
+      val importResult = importer.import(sizedInputStream(testCsv))
+
+      Assertions.assertEquals(
```
Comment (sgrimm) [inline src/test/kotlin/com/terraformation/backend/documentproducer/db/VariableImporterTest.kt]: We generally use static imports for the assertion methods in `Assertions`.

File: src/main/resources/db/migration/0250/V282__DocumentProducerTables.sql [hunk 21]
```diff
@@ -24,6 +29,12 @@ CREATE TABLE docprod.variables (
     stable_id TEXT NOT NULL,
     name TEXT NOT NULL,
     description TEXT,
+    deliverable_id BIGINT REFERENCES accelerator.deliverables ON DELETE SET NULL,
+    deliverable_question TEXT,
+    dependency_variable_stable_id TEXT,
```
Comment (sgrimm) [inline src/main/resources/db/migration/0250/V282__DocumentProducerTables.sql:34]: We'll need to keep track of the order of the questions.

File: src/test/kotlin/com/terraformation/backend/documentproducer/model/VariableTest.kt [hunk 22]
```diff
@@ -205,19 +205,19 @@ class VariableTest {
...
+    //        testValidate(
+    //            SectionVariable(baseVariable(), true),
+    //            NewSectionValue(
+    //                baseValue(),
+    //                SectionValueVariable(
+    //                    nonexistentVariableId,
+    //                    VariableUsageType.Injection,
+    //                    VariableInjectionDisplayStyle.Block)))
+    //      }
+    //    }
```
Comment (sgrimm) [inline src/test/kotlin/com/terraformation/backend/documentproducer/model/VariableTest.kt]: Might be worth a TODO here if the intent is to update this test. But it should work as-is, shouldn't it? The variable won't exist since nothing has inserted it.

**Review by sgrimm — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/terraformation/backend/admin/AdminDocumentProducerController.kt (hunk 1 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/admin/AdminDocumentProducerController.kt]: OK that makes sense, I will remove it.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 8 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: I will put it back the way you had it, much cleaner. It was lost in the shuffle when I was resolving merge conflicts.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 9 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Another change lost in the shuffle, thanks for the catch.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 11 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Good point... it seems I should remove this condition and also remove the `.and(VARIABLE_MANIFEST_ID.eq(manifestId))` condition in `fetchRecommendedBy`. Does this mean there is a missing test or a false positive?

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 10 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Removing those lines makes these two tests ([1](https://github.com/terraware/terraware-server/pull/2033/files#diff-b86832f3f41fd25f07ca39b2e5cf533b98b5b04550d2f2e8f32c75f2a3ae1361R152), [2](https://github.com/terraware/terraware-server/pull/2033/files#diff-abe3f21b9a94e358017d9e6a623c4354d4edae5bdf85d3bc3cae6ec832dfbd3dR103)) fail for some reason.

I will take a look at this after resolving the rest of your comments.

Diff looks like this for reference:
```
diff --git a/src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt b/src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt
index ecc654c3257..91050eb20ca 100644
--- a/src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt
+++ b/src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt
@@ -102,10 +102,6 @@ class VariableStore(
           .select(VARIABLE_ID)
           .from(VARIABLE_MANIFEST_ENTRIES)
           .where(VARIABLE_MANIFEST_ID.eq(manifestId))
-          .andNotExists(
-              DSL.selectOne()
-                  .from(VARIABLE_TABLE_COLUMNS)
-                  .where(VARIABLE_ID.eq(VARIABLE_TABLE_COLUMNS.VARIABLE_ID)))
           .andNotExists(
               DSL.selectOne()
                   .from(VARIABLE_SECTIONS)

```

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 11 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: Never mind, I see that it will just return an empty list if there is no manifest ID.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 13 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt]: These were needed when I had the subclasses, will put it back.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt (hunk 15 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt]: Good call out, thanks

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableImporter.kt (hunk 16 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableImporter.kt]: Not in this import!

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableImporter.kt (hunk 18 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableImporter.kt]: Not used, not needed!

File: src/test/kotlin/com/terraformation/backend/documentproducer/model/VariableTest.kt (hunk 22 above)
Comment (nickgraz) [reply src/test/kotlin/com/terraformation/backend/documentproducer/model/VariableTest.kt]: I put this back, it was related to removing `fetchVariable` as an argument to all of the variable validation functions, but I was able to put that back.

File: src/main/kotlin/com/terraformation/backend/documentproducer/model/Variable.kt (hunk 19 above)
Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/documentproducer/model/Variable.kt]: I had removed this when I thought I needed to split up the fetch contexts, I put it all back and everything is still working.

File: src/main/resources/db/migration/0250/V282__DocumentProducerTables.sql (hunk 21 above)
Comment (nickgraz) [reply src/main/resources/db/migration/0250/V282__DocumentProducerTables.sql:34]: I'm planning on adding that in another PR since this one has already ballooned quite a bit.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt (hunk 12 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/VariableStore.kt:315]: I think my comment above is still valid.

File: src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt (hunk 14 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/documentproducer/db/variable/VariableCsvValidator.kt:110]: Sanity check would still be good here.

File: src/main/kotlin/com/terraformation/backend/i18n/Messages.kt [hunk 23]
```diff
@@ -381,10 +381,47 @@ class Messages {
...
+
+  fun variableCsvDeliverableDoesNotExist() = getMessage("variablesCsvDeliverableDoesNotExist")
+
+  fun variablesCsvDependencyConfigIncomplete() =
+      getMessage("variablesCsvDependencyConfigIncomplete")
+
+  fun variablesCsvDependencyVariableStableIdDoesNotExist() =
+      getMessage("variablesCsvDependencyVariableStableIdDoesNotExist")
+
+  fun variablesCsvNameLineBreak() = getMessage("variablesCsvNameLineBreak")
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/i18n/Messages.kt:423]: This seems redundant with line 386.

File: src/test/kotlin/com/terraformation/backend/documentproducer/db/VariableImporterTest.kt [hunk 24]
```diff
@@ -0,0 +1,820 @@
...
+      val actualVariableOAL = getVariableByName("Organization Name As List")
+      val actualVariableTextOAL = variableTextsDao.fetchOneByVariableId(actualVariableOAL.id!!)
+      val actualVariablePS = getVariableByName("Prior Scenario")
+      val actualVariableTextPS = variableTextsDao.fetchOneByVariableId(actualVariablePS.id!!)
+      val actualVariablePSAL = getVariableByName("Prior Scenario As List")
+      val actualVariableTextPSAL = variableTextsDao.fetchOneByVariableId(actualVariablePSAL.id!!)
+
+      assertEquals(emptyList<String>(), importResult.errors, "no errors")
+
+      Assertions.assertFalse(actualVariableO.isList!!, "single input without list is not a list")
```
Comment (sgrimm) [inline src/test/kotlin/com/terraformation/backend/documentproducer/db/VariableImporterTest.kt:255]: Should use a static import for this and the other assert methods. When I end up with this (usually after I use IntelliJ to duplicate a class and it decides to replace the existing static imports with qualified names) I usually search for `Assertions.` and walk through the file seeing what assertions I need to import.

**Review by sgrimm — APPROVED** [review]
I think this is good enough to merge into the feature branch, though there are a few things that could stand to be changed. Up to you if you want to address them here or in a followup.

**Comment by nickgraz** [comment]
> I think this is good enough to merge into the feature branch, though there are a few things that could stand to be changed. Up to you if you want to address them here or in a followup.

Thanks @sgrimm , I'll merge it in and address these comments in another PR
